### PR TITLE
Popover legend fixes

### PR DESCRIPTION
**Pull request by JakeSidSmith**
Relies on: https://github.com/Opsimize/c3/pull/12

* Take into account Firefox scrolling for popover legends
* Ensure popover legends are removed on destroy

### CODE REVIEW

File: c3.js [hunk 1]
```diff
@@ -2010,7 +2012,16 @@
             var parent = d3.select(__bindto);
             var parentRect = parent[0][0].getBoundingClientRect();
 
-            tooltipTop = Math.max(tooltipTop, -100) + parentRect.top + document.body.scrollTop;
+            var scrollTop = document.body.scrollTop;
+            var html = document.getElementsByTagName('html')[0];
```
Comment (JakeSidSmith) [inline c3.js]: Firefox uses `html.scrollTop` rather than `document.body`.

File: c3.js [hunk 2]
```diff
@@ -2010,7 +2012,16 @@
...
+
+            if (html && html.scrollTop) {
+              scrollTop = html.scrollTop
+            }
+
+            scrollTop = scrollTop || 0;
```
Comment (JakeSidSmith) [inline c3.js]: This ensures that we wont have an `undefined` value.

File: c3.js [hunk 3]
```diff
@@ -76,6 +76,8 @@
      */
     c3.generate = function (config) {
 
+        var tooltip;
```
Comment (JakeSidSmith) [inline c3.js]: We make `tooltip` a global so it can be easily removed on destroy.

File: c3.js [hunk 4]
```diff
@@ -4014,8 +4025,14 @@
             withTransition = isDefined(options.withTransition) ? options.withTransition : true;
             withTransitionForTransform = isDefined(options.withTransitionForTransform) ? options.withTransitionForTransform : true;
 
-            function updatePositions(textElement, id, reset) {
-                var box = textElement.getBoundingClientRect(),
+            function updatePositions(textElement, id, reset, textHolder) {
+                var textClone = textHolder.append(function () {
+                        return textElement.cloneNode(true);
```
Comment (JakeSidSmith) [inline c3.js]: Create a clone of the text element for each legend item. These are appended to an SVG on the body.

File: c3.js [hunk 5]
```diff
@@ -4014,8 +4025,14 @@
...
+                    });
+
+                textClone.attr('style', 'visibility:hidden;');
+
+                var box = textClone[0][0].getBBox ? textClone[0][0].getBBox() : textClone[0][0].getBoundingClientRect(),
```
Comment (JakeSidSmith) [inline c3.js]: Check if exists & use `getBBox` (as Firefox cannot measure SVG elements with getBoundingClientRect).

File: c3.js [hunk 6]
```diff
@@ -4114,9 +4131,18 @@
                         __legend_item_onmouseout(id);
                     }
                 });
+
+            var docBody = d3.select('body');
+
+            var textHolder = docBody.append('svg');
```
Comment (JakeSidSmith) [inline c3.js]: Create an SVG to append our legend texts to, and stick it in the body.

File: c3.js [hunk 7]
```diff
@@ -4114,9 +4131,18 @@
...
+
+            textHolder.attr('width', 0)
+                .attr('height', 0)
+                .attr('style', 'position:absolute;overflow:hidden;');
+
             l.append('text')
                 .text(function (id) { return isDefined(__data_names[id]) ? __data_names[id] : id; })
-                .each(function (id, i) { updatePositions(this, id, i === 0); })
+                .each(function (id, i) { updatePositions(this, id, i === 0, textHolder); })
```
Comment (JakeSidSmith) [inline c3.js]: This SVG gets passed into `updatePositions` so we can append the texts to it for measuring,

File: c3.js [hunk 8]
```diff
@@ -4139,11 +4165,14 @@
             legend.selectAll('text')
                 .data(targetIds)
                 .text(function (id) { return isDefined(__data_names[id]) ? __data_names[id] : id; }) // MEMO: needed for update
-                .each(function (id, i) { updatePositions(this, id, i === 0); })
+                .each(function (id, i) { updatePositions(this, id, i === 0, textHolder); })
               .transition().duration(withTransition ? 250 : 0)
                 .attr('x', xForLegendText)
                 .attr('y', yForLegendText);
 
+            textHolder.remove();
```
Comment (JakeSidSmith) [inline c3.js]: Remove our temp SVG element.

File: c3.js [hunk 9]
```diff
@@ -4557,6 +4586,9 @@
         c3.destroy = function () {
             c3.data.targets = undefined;
             c3.data.xs = {};
+            if (tooltip && typeof tooltip.remove === 'function') {
+              tooltip.remove();
```
Comment (JakeSidSmith) [inline c3.js]: If the popover legend exists & can be removed... remove it.

**Comment by SamRichards** [comment]
:sparkles:

**Comment by CHR15-** [comment]
:sparkles:

**Pull request merged**